File: server/src/core/movement.js
/**
 * 移動の確定 — docs/design/09-protocol.md §5.1
 *
 * クライアントは「進みたい向き」しか送れない。座標は送れない。
 * サーバがここで速度上限と壁を検証して確定させる。
 * 偽装できるのは向きだけで、それはこの関数で潰れる。
 */

/** タイル/秒 */
export const SPEED = 4.2;
/** RTT のゆらぎを吸収する許容係数 */
export const TOLERANCE = 1.2;

/**
 * @param {import('./grid.js').Grid} grid
 * @param {{x:number,y:number,dir:number}} pos  現在位置（この関数は変更しない）
 * @param {number} dx  -1..1（範囲外は丸める）
 * @param {number} dy  -1..1
 * @param {number} dtSec  経過秒。負や過大な値は丸める
 * @returns {{x:number,y:number,dir:number,moved:boolean,clamped:boolean}}
 */
export function resolveMove(grid, pos, dx, dy, dtSec) {
  // 入力の正規化。壊れた値でサーバが壊れないようにする
  dx = clampUnit(dx);
  dy = clampUnit(dy);
  const dt = Math.min(Math.max(dtSec, 0), 0.25); // 250ms を超える一括移動は認めない

  const len = Math.hypot(dx, dy);
  if (len === 0 || dt === 0) {
    return { x: pos.x, y: pos.y, dir: pos.dir, moved: false, clamped: false };
  }

  const want = SPEED * dt;
  const max = SPEED * dt * TOLERANCE;
  const step = Math.min(want, max);

  const nx = pos.x + (dx / len) * step;
  const ny = pos.y + (dy / len) * step;

  // 軸ごとに判定する（壁ずり）。斜めに詰まって止まらないようにするため
  let x = pos.x, y = pos.y, clamped = false;
  if (!grid.isBlocked(nx, pos.y)) x = nx; else clamped = true;
  if (!grid.isBlocked(x, ny)) y = ny; else clamped = true;

  const dir = Math.atan2(dx, dy);
  const moved = x !== pos.x || y !== pos.y;
  return { x, y, dir, moved, clamped };
}

/**
 * クライアントが主張した座標が、直前の確定座標から到達可能かを検証する。
 * 通常の経路では使わない（座標を受け取らないため）が、
 * 将来クライアントが座標を送る経路を足したときの門番として置いておく。
 */
export function isReachable(prev, next, dtSec) {
  const max = SPEED * Math.min(Math.max(dtSec, 0), 0.25) * TOLERANCE;
  return Math.hypot(next.x - prev.x, next.y - prev.y) <= max + 1e-9;
}

function clampUnit(v) {
  if (!Number.isFinite(v)) return 0;
  return Math.min(1, Math.max(-1, v));
}

File: unity/Assets/Hidamari/Core/Movement.cs
// 移動の確定 — docs/design/09-protocol.md §5.1
//
// ★ server/src/core/movement.js の写し。UnityEngine を参照しない。
//   クライアントは「進みたい向き」しか送れない。座標は送れない。
//   ここで確定させるのは**予測**であって、権威はサーバにある。
//   だからこそ、サーバと1三角形の狂いも無く同じでなければならない。
//
// ★ 写し元と一致していることは shared/golden/movement.tsv で見張る。
//   `dotnet run --project unity/Tests.Headless` か、Unity の EditMode テスト。
//
// ★ double で計算する。JS の Number は double なので、float にすると表と合わない。

namespace Hidamari.Core
{
    public struct MoveResult
    {
        public double X, Y, Dir;
        public bool Moved, Clamped;
    }

    public static class Movement
    {
        /// タイル/秒
        public const double Speed = 4.2;
        /// RTT のゆらぎを吸収する許容係数
        public const double Tolerance = 1.2;
        /// 一括移動を認めない上限（秒）
        public const double MaxStepSec = 0.25;

        public static MoveResult Resolve(Grid grid, double x, double y, double dir,
                                         double dx, double dy, double dtSec)
        {
            // 入力の正規化。壊れた値でクライアントが壊れないようにする
            dx = ClampUnit(dx);
            dy = ClampUnit(dy);
            var dt = System.Math.Min(System.Math.Max(dtSec, 0), MaxStepSec);

            var len = System.Math.Sqrt(dx * dx + dy * dy);
            if (len == 0 || dt == 0)
                return new MoveResult { X = x, Y = y, Dir = dir, Moved = false, Clamped = false };

            // want と max は同じ式から出るので step は常に want。
            // 写し元のまま残してある（許容係数を上げたくなったときここを触る）
            var want = Speed * dt;
            var max = Speed * dt * Tolerance;
            var step = System.Math.Min(want, max);

            var nx = x + (dx / len) * step;
            var ny = y + (dy / len) * step;

            // 軸ごとに判定する（壁ずり）。斜めに詰まって止まらないようにするため
            double ox = x, oy = y;
            var clamped = false;
            if (!grid.IsBlocked(nx, y)) ox = nx; else clamped = true;
            if (!grid.IsBlocked(ox, ny)) oy = ny; else clamped = true;

            return new MoveResult
            {
                X = ox,
                Y = oy,
                // ★ Atan2 の引数の順が JS と同じ (dx, dy) であること。
                //   C# の癖で (dy, dx) と書くと、向きだけが 90 度ずれる
                Dir = System.Math.Atan2(dx, dy),
                Moved = ox != x || oy != y,
                Clamped = clamped,
            };
        }

        /// クライアントが主張した座標が、直前の確定座標から到達可能か
        public static bool IsReachable(double px, double py, double nx, double ny, double dtSec)
        {
            var max = Speed * System.Math.Min(System.Math.Max(dtSec, 0), MaxStepSec) * Tolerance;
            var d = System.Math.Sqrt((nx - px) * (nx - px) + (ny - py) * (ny - py));
            return d <= max + 1e-9;
        }

        static double ClampUnit(double v)
        {
            // JS の Number.isFinite と同じ扱い。NaN も Infinity も 0 にする
            if (double.IsNaN(v) || double.IsInfinity(v)) return 0;
            return System.Math.Min(1, System.Math.Max(-1, v));
        }
    }
}
